File: 05.py
import time
from libs import *


def main():
    print("--- Running Day 5 ---")

    start = time.perf_counter()

    try:
        with open("05.txt", "r") as f:
            data = f.read()
    except Exception as e:
        raise ValueError("Unable to read file") from e

    stack_input, moves = data.split("\n\n")

    input = stack_input.splitlines()
    input.reverse()
    stack_len = [int(x) for x in input[0].split()].pop()

    stacks_p1 = [[] for _ in range(0, stack_len)]
    stacks_p2 = [[] for _ in range(0, stack_len)]
    for line in input[1:]:
        for i, c in enumerate(line[1::4]):
            if not c.isspace():
                stacks_p1[i].append(c)
                stacks_p2[i].append(c)

    for line in moves.splitlines():
        _, num_s, _, init_s, _, dest_s = line.split()
        num, init, dest = int(num_s), int(init_s), int(dest_s)

        str = []
        for _ in range(num):
            stacks_p1[dest - 1].append(stacks_p1[init - 1].pop())
            str.append(stacks_p2[init - 1].pop())
        str.reverse()
        stacks_p2[dest - 1].extend(str)

    answer_p1 = ""
    for x in range(stack_len):
        if len(stacks_p1[x]) > 0:
            answer_p1 += stacks_p1[x][-1]

    print(f"Output Part 1: {answer_p1}")

    answer_p2 = ""
    for x in range(stack_len):
        if len(stacks_p2[x]) > 0:
            answer_p2 += stacks_p2[x][-1]

    print(f"Output Part 2: {answer_p2}")

    end = time.perf_counter()
    rtime = (end - start) * 1000  # sec -> ms
    print(f"Took {format_runtime(rtime)}\n")


if __name__ == "__main__":
    main()

File: 05.js
const fs = require('fs');
const { performance } = require('perf_hooks');
const { formatRuntime } = require('./libs');

async function readInput() {
  const input = await fs.promises.readFile('05.txt', 'utf-8');
  return input.trim().split('\n\n');
}

async function puzzle() {
  console.log('--- Running Day 5 ---');

  const start = performance.now();

  const [stackInput, moves] = await readInput();

  let input = stackInput.split('\n');
  input = input.reverse();

  const stackLen = Number(
    input[0]
      .split(/(\s+)/)
      .filter((e) => e.trim().length > 0)
      .pop()
  );

  const stacksP1 = [];
  const stacksP2 = [];
  for (let i = 0; i < stackLen; i += 1) {
    stacksP1[i] = [];
    stacksP2[i] = [];
  }

  for (let i = 1; i < input.length; i += 1) {
    const line = input[i];
    let index = 0;
    for (let j = 1; j <= line.length; j += 4) {
      if (line.slice(j, j + 1) !== '') {
        if (line.slice(j, j + 1) !== ' ') {
          stacksP1[index].push(line.slice(j, j + 1));
          stacksP2[index].push(line.slice(j, j + 1));
        }
        index += 1;
      }
    }
  }

  moves.split('\n').forEach((line) => {
    const strs = line.split(' ');
    const num = +strs[1];
    const init = +strs[3];
    const dest = +strs[5];

    let letters = '';
    for (let i = 0; i < num; i += 1) {
      const letter = stacksP1[init - 1].pop();
      if (letter) {
        stacksP1[dest - 1].push(letter);
      }

      const letterP2 = stacksP2[init - 1].pop();
      if (letterP2) {
        letters += letterP2;
      }
    }
    letters = letters.split('').reverse().join('');
    stacksP2[dest - 1].push(...letters.split(''));
  });

  let answerP1 = '';
  for (let x = 0; x < stackLen; x += 1) {
    if (stacksP1[x].length > 0) {
      answerP1 += stacksP1[x].pop();
    }
  }

  console.log('Part 1: ' + answerP1);

  let answerP2 = '';
  for (let x = 0; x < stackLen; x += 1) {
    if (stacksP2[x].length > 0) {
      answerP2 += stacksP2[x].pop();
    }
  }

  console.log('Part 2: ' + answerP2);

  const end = performance.now();
  const rtime = end - start;
  console.log('Took:', formatRuntime(rtime));
}

puzzle();
